rust_src/src/terminal.rs

```diff
@@ -0,0 +1,83 @@
+use libc::c_int;
+use remacs_sys::{frame, glyph};
+
+#[no_mangle]
+pub unsafe extern "C" fn update_begin(f: *mut frame) {
```

Comment:
We can use `LispFrameRef` defined in `frames.rs`. The code can be like this.
```rust
#[no_mangle]
pub extern "C" fn update_begin(f: LispFrameRef) {
   if let Some(hook) = (*(f.terminal)).update_begin_hook {
        hook(f.as_mut())
    }
}
```

if we defined a new `TerminalRef` type just as the `LispFrameRef`, the above code could be more rusty. All the magic are from `ExternalPtr` in lisp.rs. You can check out it.

Reply:
`*mut window` is just same.

Reply:
Same with `*mut window`.

Reply:
You can copy it from #871

Reply:
@harryfei this is a C exposed function not a `lisp_fn`.

Reply:
Is it safe to use that here? It's `#[repr(C)]`, not `#[repr(transparent)]`.

Reply:
@bb010g look at https://github.com/Wilfred/remacs/blob/a613f9ff65ad7ef039c26c403302e2f7e4b5410f/rust_src/src/windows.rs#L652 then look at `LispWindowRef`